### PR TITLE
✨ CDK: availability check - handle HttpErrors which happen during slice extraction 

### PR DESCRIPTION
**Pull request by midavadim**
Problem:
CDK does not catch '403' HttpError coming from slice extraction during the availability check
CDK does not properly log another HttpErrors which happens during the availability check

## What
for availability check - handle  HttError happens during slice extraction (reading of parent stream), 
updated reason messages,
moved check availability call under common try/except which handles errors during usual stream read, 
moved log messages which indicate start of the stream sync before availability check in to make to understand which stream is the source of errors

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py [hunk 1]
```diff
@@ -109,12 +109,12 @@ def read(
                         f"The requested stream {configured_stream.stream.name} was not found in the source."
                         f" Available streams: {stream_instances.keys()}"
                     )
-                stream_is_available, error = stream_instance.check_availability(logger, self)
-                if not stream_is_available:
-                    logger.warning(f"Skipped syncing stream '{stream_instance.name}' because it was unavailable. Error: {error}")
-                    continue
                 try:
                     timer.start_event(f"Syncing stream {configured_stream.stream.name}")
+                    stream_is_available, reason = stream_instance.check_availability(logger, self)
```
Comment (midavadim) [inline airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py:114]: moved check availability call under common try/except which handles errors during usual stream read,
moved log messages which indicate start of the stream sync before availability check in to make to understand which stream is the source of errors

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/availability_strategy.py [hunk 2]
```diff
@@ -95,17 +106,16 @@ def reasons_for_unavailable_status_codes(
         why 'status code' may have occurred and how the user can resolve that
         error, if applicable.
         """
-        forbidden_error_message = f"The endpoint to access stream '{stream.name}' returned 403: Forbidden. "
-        forbidden_error_message += "This is most likely due to insufficient permissions on the credentials in use. "
-        forbidden_error_message += self._visit_docs_message(logger, source)
-
-        reasons_for_codes: Dict[int, str] = {requests.codes.FORBIDDEN: forbidden_error_message}
+        reasons_for_codes: Dict[int, str] = {
```
Comment (midavadim) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/availability_strategy.py:109]: updated reason message with todo action. 
re-arranged code a bit

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/availability_strategy.py [hunk 3]
```diff
@@ -40,6 +40,11 @@ def check_availability(self, stream: Stream, logger: logging.Logger, source: Opt
             # without accounting for the case in which the parent stream is empty.
             reason = f"Cannot attempt to connect to stream {stream.name} - no stream slices were found, likely because the parent stream is empty."
             return False, reason
+        except HTTPError as error:
```
Comment (midavadim) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/availability_strategy.py:43]: handle  HttError happens during slice extraction (reading of parent stream),

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py [hunk 4]
```diff
@@ -41,6 +41,7 @@ def check_connection(self, source: Source, logger: logging.Logger, config: Mappi
             try:
                 stream_is_available, reason = availability_strategy.check_availability(stream, logger, source)
                 if stream_is_available:
+                    # why do we return here and not try next stream?
```
Comment (midavadim) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py:44]: is it a bug or intentional behavior?

**Review by midavadim — COMMENTED** [review]
see my annotations

**Review by bazarnov — APPROVED** [review]

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py (hunk 4 above)
Comment (erohmensing) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py:44]: This looks like a bug to me! Can you add a test which has 2 streams, for which the first is available and the second is not?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/availability_strategy.py [hunk 5]
```diff
@@ -95,17 +106,16 @@ def reasons_for_unavailable_status_codes(
...
         """
-        forbidden_error_message = f"The endpoint to access stream '{stream.name}' returned 403: Forbidden. "
-        forbidden_error_message += "This is most likely due to insufficient permissions on the credentials in use. "
-        forbidden_error_message += self._visit_docs_message(logger, source)
-
-        reasons_for_codes: Dict[int, str] = {requests.codes.FORBIDDEN: forbidden_error_message}
+        reasons_for_codes: Dict[int, str] = {
+            requests.codes.FORBIDDEN: "This is most likely due to insufficient permissions on the credentials in use. "
+            "Try to grant required permissions/scopes or re-authenticate"
+        }
```
Comment (erohmensing) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/availability_strategy.py:112]: Why did we remove the reference to the docs?

**Comment by erohmensing** [comment]
Thanks for looking into this! Noted that I do think you found a bug re: list of streams to check - I'll slot a full review in later this week when I have more time

**Comment by sherifnada** [comment]
adding in @brianjlai to represent extensibility for a review

**Review by brianjlai — COMMENTED** [review]
overall looks good, there are some discussion points to implement, but once done will 👍 from extensibility team side

**Review by lazebnyi — APPROVED** [review]
LGTM!

**Review by midavadim — COMMENTED** [review]
Updated review

**Review by brianjlai — APPROVED** [review]

**Comment by sherifnada** [comment]
@erohmensing @girarda @pedroslopez do you want your review to be blocking?

**Review by girarda — APPROVED** [review]
lgtm! 🚢

**Comment by midavadim** [comment]
@erohmensing @pedroslopez could you please finish this review when you have spare time.

Thank you

**Review by pedroslopez — APPROVED** [review]

**Pull request merged**